kcidb/test_oo.py

```diff
@@ -203,3 +204,225 @@ def test_two_object(self):
             '>checkout#>build#>test#',
             {"test", "build", "checkout", "revision"}, 2
         )
+
+
+class KCIDBTraversingTestCase(kcidb.unittest.TestCase):
+    """Test case for traversing objects."""
+
+    # Conform to unittest conventions, pylint: disable=invalid-name
+
+    def query_str(self, pattern_string):
+        """Run OO client query with a parsed string."""
+        return self.client.query(
+            kcidb.orm.Pattern.parse(pattern_string)
+        )
+
+    def assertNumberOfInstances(self, oo_object_list, num_instances):
+        """
+        Check that the correct number of instances is returned for
+        the object list.
+        Args:
+            oo_object_list: A list of kcidb.oo.Object instances.
+            num_instances:  The number of instance of object contained
+                            in the oo_object_list.
+        """
+        assert isinstance(oo_object_list, list)
+        assert isinstance(num_instances, int)
+
+        self.assertEqual(
+            len(oo_object_list), num_instances,
+            msg=f"Invalid number of {oo_object_list[0]} instances"
+        )
+
+    def assertDictionaryContains(self, oo_dict, item):
+        """
+        Check that the dictionary returned is indexed by the correct
+        indices, and the total quantity is right.
+        Args:
+            oo_dict:    A dictionary of objects indexed by two indices.
+            item:       A list of first indexing items.
+       """
+        for key in oo_dict.keys():
+            self.assertIn(key, item)
+            self.assertEqual(oo_dict[key].keys(), item[key].keys())
+            for type_objs in oo_dict[key]:
+                self.assertEqual(
+                    len(oo_dict[key][type_objs]), item[key][type_objs]
+                )
```

Comment:
Having and using this kind of function for our tests seems like a good idea.

However, it needs better documentation, explaining what exactly the last argument should be. It can also be made more elegant and more universal with recursion, allowing us to validate any number of nested dictionaries, or even arrays. We also need to verify that none of them contain anything more than what we expect.

We can shorten `Dictionary` to just `Dict` in the name, and we also need to be explicit that it's not checking actual contents, but only the counts. A nice addition could be checking the types of objects too, but we can start with just the counts.

Can you figure out how to do that?

Reply:
Yes. of course. Check out how I went about it.

Reply:
Yes. of course. Check out how I went about it.